tiatoolbox/dataloader/wsireader.py

```diff
@@ -82,39 +91,524 @@ def slide_info(self):
 
         Returns:
-            ndarray : image array
+            ndarray : array of size MxNx3
+            M=end_h-start_h, N=end_w-start_w
 
+        Examples:
+            >>> from tiatoolbox.dataloader import wsireader
+            >>> from matplotlib import pyplot as plt
+            >>> wsi_obj = wsireader.WSIReader("/path/to/a/wsi")
+            >>> # Read a region at level 0 (baseline / full resolution)
+            >>> bounds = [1000, 2000, 2000, 3000]
+            >>> img = wsi_obj.read_bounds(*bounds)
+            >>> plt.imshow(img)
+            >>> # This could also be written more verbosely as follows
+            >>> img = wsi_obj.read_bounds(
+            ...     start_w=bounds[0],
+            ...     start_h=bounds[1],
+            ...     end_w=bounds[2],
+            ...     end_h=bounds[3],
+            ...     resolution=0,
+            ...     units="level",
+            ... )
+            >>> plt.imshow(img)
+
+        Note: The field of view remains the same as resolution is varied.
+
+        .. figure:: images/read_bounds_tissue.png
+            :width: 512
+            :alt: Diagram illustrating read_bounds
+
+        This is because
+        the bounds are in the baseline (level 0) reference
+        frame. Therefore, varying the resolution does not change what is
+        visible within the output image.
+
+        If the WSI does not have a resolution layer
+        corresponding exactly to the requested resolution
+        (shown above in white with a dashed outline), a larger
+        resolution is downscaled to achieve the correct requested output
+        resolution.
+
+        If the requested resolution is higher than the
+        baseline (maximum resultion of the image), then bicubic
+        interpolation is applied to the output image.
         """
-        raise NotImplementedError
+        raise NotImplementedError()
+
+    def read_region(self, location, level, size):
+        """Read a region of the whole slide image (OpenSlide format args).
+
+        This function is to help with writing code which is backwards
+        compatible with OpenSlide. As such, it has the same arguments.
+
+        This interally calls :func:`read_rect` which should be
+        implemented by any :class:`WSIReader` subclass.
+        Therefore, some WSI formats which
+        are not supported by OpenSlide, such as Omnyx JP2 files, may
+        also be readable with the same syntax.
+
+        Args:
+            location: (x, y) tuple giving the top left pixel in the
+                level 0 reference frame.
+            level: the level number.
+            size: (width, height) tuple giving the region size.
+
+        Returns:
+            PIL.Image: Image containing the contents of the region.
+        """
+        return self.read_rect(
+            location=location, size=size, resolution=level, units="level"
+        )
+
+    def slide_thumbnail(self, resolution=1.25, units="power"):
```

Comment:
get_thumbnail?

Reply:
Happy to change in this PR if you like.

Reply:
Done in commit https://github.com/TIA-Lab/tiatoolbox/pull/40/commits/674d0851d6cf6b9c8e3e431ed5aa269d360c199e